clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -788,23 +767,28 @@ public void transitionToStale() {
     }
 
     /**
-     * Reconcile the assignment that has been received from the server and for which topic names
-     * are resolved, kept in the {@link #assignmentReadyToReconcile}. This will commit if needed,
-     * trigger the callbacks and update the subscription state. Note that only one reconciliation
+     * Reconcile the assignment that has been received from the server. If for some topics, the
+     * topic ID cannot be matched to a topic name, a metadata update will be triggered and only
+     * the subset of topics that are resolvable will be reconciled. Reconciliation will trigger the
+     * callbacks and update the subscription state. Note that only one reconciliation
      * can be in progress at a time. If there is already another one in progress when this is
      * triggered, it will be no-op, and the assignment will be reconciled on the next
      * reconciliation loop.
      */
-    boolean reconcile() {
+    void reconcile() {
+        if (targetAssignmentReconciled()) {
+            log.debug("Ignoring reconciliation attempt. Target assignment is equal to the " +
+                    "current assignment.");
+            return;
+        }
         if (reconciliationInProgress) {
             log.debug("Ignoring reconciliation attempt. Another reconciliation is already in progress. Assignment " +
-                    assignmentReadyToReconcile + " will be handled in the next reconciliation loop.");
-            return false;
+                currentTargetAssignment + " will be handled in the next reconciliation loop.");
+            return;
         }
 
-        // Make copy of the assignment to reconcile as it could change as new assignments or metadata updates are received
-        SortedSet<TopicIdPartition> assignedTopicIdPartitions = new TreeSet<>(TOPIC_ID_PARTITION_COMPARATOR);
-        assignedTopicIdPartitions.addAll(assignmentReadyToReconcile);
+        // Resolve metadata for target assignment
+        SortedSet<TopicIdPartition> assignedTopicIdPartitions = resolveMetadataForTargetAssignment();
```

Comment:
nit: The name of the method feels a bit weird. Ok, it indeed triggers the metadata request but it returns the subset of the target assignment with metadata. I wonder if we could find a better name.

Reply:
Done